File: src/main/java/alg03sorting/S5BottomUpMergeSort.java
package alg03sorting;


/**
 * Divide and conquer (split-sort-merge) algorithm.
 * The execution time is as fast as O(N*log2(N)).
 * <p>
 * It is the same mergesort, only without the recursion.
 */
public class S5BottomUpMergeSort extends S4MergeSort {
    public <T extends Comparable<T>> void sort(T[] array) {
        int N = array.length;
        Object[] auxArray = new Object[N];
        for (int size = 1; size < N; size = size + size) {
            for (int lo = 0; lo < N - size; lo += size + size) {
                merge(array, auxArray, lo, lo + size - 1, Math.min(lo + size + size - 1, N - 1));
            }
        }
    }
}

File: src/alg03sorting/S5BottomUpMergeSort.ts
import {S4MergeSort} from "./S4MergeSort";
import {S0Sort} from "./S0Sort";

/**
 * It is the same merge sort, only without the recursion.
 */
export class S5BottomUpMergeSort<T> extends S4MergeSort<T> implements S0Sort<T> {
    sort(array: T[]): void {
        const auxArray = Array<T>(array.length);
        for (let size = 1; size < array.length; size = 2 * size) {
            for (let leftThreshold = 0; leftThreshold < array.length - size; leftThreshold += 2 * size) {
                const midThreshold = leftThreshold + size - 1;
                const rightThreshold = Math.min(leftThreshold + 2 * size - 1, array.length - 1);
                this.merge(array, auxArray, leftThreshold, midThreshold, rightThreshold);
            }
        }
    }
}
